drivers: counter: introduce counter capture api

XenuIsWatching opened the pull request:
This introduces apis for input capture on the STM32. This is something
agnostic enough that is seen on other vendors such as Atmel's SAM line
and others.

A use case for this would be time stamping a rising and/or falling edge
where precision and lack of jitter is important. The only way to currently
do this in zephyr is to set up a GPIO interrupt and then read the counter
or kernel time... but this can have limitations of software such as the
processor may have a lot going on and can be subject to interrupt latencies.
Having a hardware 'latch' done of the rising and/or falling edge with
respect to a timer can eliminate these imprecisions.

How this is to be used, is that an application is to call
`capture_callback_set` with the configuration of the channel number, flags
(such as falling, rising, or both edges), and then a callback function. Then
the `capture_enable` function is to be called to enable this. The callback
has the timestamp of the edge in ticks and the edge that it captured (rising
or falling).

----

~~This also introduces an api for set the counter time.~~

~~Also, this set frequency to be 64 bits rather than 32 bits, as counters, 
such as those that use fractional adders, tick at resolutions
greater than 4294967295Hz. This changes the freq in the common info
to a uint64_t. This retains the existing `counter_get_freq` function which
will still return a 32bit value. This also adds a `counter_get_freq_64`
function to get a 64bit value for the frequency.~~

~~Add inline helper functions for converting ticks to nanoseconds.~~

~~Also implement an inline helper for converting microseconds or
nanoseconds with a 64bit tick value.~~

split in to https://github.com/zephyrproject-rtos/zephyr/pull/94189

henrikbrixandersen commented:
How is this functionally different from what is offered by the PWM capture API?

XenuIsWatching commented:
> How is this functionally different from what is offered by the PWM capture API?

That PWM capture measures the period and duty cycle of a pulse. This counter capture captures the exact time of a rising and/or falling edge when it happens.

nordic-krch commented on include/zephyr/drivers/counter.h:
```diff
@@ -231,13 +264,25 @@ typedef int (*counter_api_set_guard_period)(const struct device *dev,
 						uint32_t flags);
 typedef uint32_t (*counter_api_get_freq)(const struct device *dev);
+typedef uint64_t (*counter_api_get_freq_64)(const struct device *dev);
+typedef int (*counter_api_capture_callback_set)(const struct device *dev,
```
I am not sure if I understand correctly but this API assumes that:
- it captures pin event (what about other possible capture sources?)
- there is a known pin associated with the channel (is it setup in DT or fixed in the peripheral?)

At least for Nordic peripheral capture can be used with any pin and any event in the system (e.g. end of SPI transfer, start of UART RX, etc.). It might be too generic to have "any event" approach so it can be narrowed down to pin only but then it would be good to include pin configuration.

Maybe similar to `struct counter_alarm_cfg`:
```
struct counter_captrure_cfg {
   gpio_dt_spec gpio;
   uint32_t flags; //may contain flags like one_shot vs all pin events
   counter_capture_cb_t cb;
   void *user_data;
}
int counter_set_channel_capture(const struct device *dev, uint8_t chan_id,
                                  const struct counter_capture_cfg *cfg);
int counter_cancel_channel_capture(const struct device *dev, uint8_t chan_id);
uint32_t counter_get_channel_capture(const struct device *dev, uint8_t chan_id);
uint64_t counter_get_channel_capture_64(const struct device *dev, uint8_t chan_id);
```
Callback can be null as there would be also an option to read the captured value using `get_capture()` (it is not always needed to get the interrupt on capture).

carlescufi commented:
Architecture WG meeting:

- Most of the conversation revolves around the fact that the API here relies on pinmux/pinctrl to cofigure the trigger, which means that the application needs to set up via those the trigger/signal that will then make the timer capture. But different vendors use different mechanisms to configure the trigger for capturing, which raises the question of whether we should address https://github.com/zephyrproject-rtos/zephyr/issues/88660 first
- @decsny suggests taking a look at the mux bindings in Linux to see if we can reuse some of this: https://www.kernel.org/doc/Documentation/devicetree/bindings/mux/mux-controller.yaml and  https://www.kernel.org/doc/Documentation/devicetree/bindings/mux/mux-consumer.yaml
- @henrikbrixandersen likes the idea and would like to collaborate to move it forward

XenuIsWatching commented:
> Architecture WG meeting:
> 
> * Most of the conversation revolves around the fact that the API here relies on pinmux/pinctrl to cofigure the trigger, which means that the application needs to set up via those the trigger/signal that will then make the timer capture. But different vendors use different mechanisms to configure the trigger for capturing, which raises the question of whether we should address [drivers: Peripheral to peripheral interconnect API #88660](https://github.com/zephyrproject-rtos/zephyr/issues/88660) first

I'm not sure that is totally blocking this implementation, likely that would only need to be implemented for the Nordic implementation, and if there are any needs of gaps found in this api I'm sure as this could be 'experimental' I'm sure those changes (if any) could come when that comes, but I'm happy to mostly certainly consider those now, I'm just not sure what those would need to be.

> * @decsny suggests taking a look at the mux bindings in Linux to see if we can reuse some of this: https://www.kernel.org/doc/Documentation/devicetree/bindings/mux/mux-controller.yaml and  https://www.kernel.org/doc/Documentation/devicetree/bindings/mux/mux-consumer.yaml

Thinking more about this, I'm not sure where this fits in, as the pinctrl is used in this implementation to select the mux on the "pad" to be routed to the counter.

> * @henrikbrixandersen likes the idea and would like to collaborate to move it forward

@henrikbrixandersen Sounds great!!! We can discuss your comments and proposes over Discord if thats best.

------

On other note :musical_note: , I'm thinking about splitting this PR, there appears to be a lot more discussion on the 'input capture' part of this... and the other changes I had to make such as changing some `uint32_t` to a `uint64_t` as well as adding some additional simple api and inline line helpers there is not much discussion on and could easily be separated from this counter capture api work.

decsny commented:
> Thinking more about this, I'm not sure where this fits in, as the pinctrl is used in this implementation to select the mux on the "pad" to be routed to the counter.

The point is related to the fact that other people brought up this recurring topic about on chip triggers. There intention was to suggest expanding the scope of this implementation to consider more than just external ios as capture triggers.

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -302,6 +345,80 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
+	return api->capture_disable(dev, chan);
+}
+#endif
+
```
Could you remove this extra empty line?

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -302,6 +345,80 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
 
 	return api->get_freq ? api->get_freq(dev) : config->freq;
 }
+#if defined(CONFIG_COUNTER_CAPTURE) || defined(__DOXYGEN__)
```
Could you add an empty line above?

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -302,6 +345,80 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
 	return api->get_freq ? api->get_freq(dev) : config->freq;
 }
+#if defined(CONFIG_COUNTER_CAPTURE) || defined(__DOXYGEN__)
+__syscall int counter_capture_callback_set(const struct device *dev,
```
Suggestion:
```c
/**
 * @brief Set callback function for a counter timestamp capture
 *
 * @param dev  Pointer to the device structure for the driver instance
 * @param chan Channel ID
 * @param flags Configuration flags (COUNTER_CAPTURE_*)
 * @param cb Callback function reference
 * @param user_data Argument passed to the callback function
 *
 * @retval 0 If successful.
 * @retval Negative error code on failure
 */

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -231,6 +260,15 @@ typedef int (*counter_api_set_guard_period)(const struct device *dev,
+						uint32_t flags,
+						counter_capture_cb_t cb,
+						void *user_data);
+typedef int (*counter_api_capture_enable)(const struct device *dev,
+					  uint8_t chan);
+typedef int (*counter_api_capture_disable)(const struct device *dev,
+					   uint8_t chan);
```
Nitpicking: fit is single lines:
```c
typedef int (*counter_api_capture_enable)(const struct device *dev, uint8_t chan);
typedef int (*counter_api_capture_disable)(const struct device *dev, uint8_t chan);

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -655,6 +772,8 @@ void counter_stm32_irq_handler(const struct device *dev)
 	BUILD_ASSERT(NUM_CH(TIM(idx)) <= TIMER_MAX_CH,				  \
 		     "TIMER too many channels");				  \
 										  \
+	IF_ENABLED(CONFIG_COUNTER_CAPTURE,					  \
+		(PINCTRL_DT_INST_DEFINE(idx);))					  \
```
```suggestion
	IF_ENABLED(CONFIG_COUNTER_CAPTURE, (PINCTRL_DT_INST_DEFINE(idx);))	  \
										  \
```
(with an escaped empty line below)

etienne-lms reviewed (commented):
Minor comments

bjarki-andreasen commented:
To get this going, we could agree to a similar approach to what we do for comparators, let vendors define the capture inputs in whatever way makes sense using the devicetree, where its all vendor specific and flexible, and add vendor specific APIs for runtime configuration of mux or other hardware options if required.

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -465,29 +486,122 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
 	chdata = &config->ch_data[id];
-	cb = chdata->callback;
-	chdata->callback = NULL;
+#ifdef CONFIG_COUNTER_CAPTURE
```
An empty line above is welcome.

XenuIsWatching replied:
done

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -558,28 +579,119 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+			}
+
+			cb(dev, id, flags, cc_val, chdata->user_data);
+		}
+	} else {
```
I seems a matter of taste. At you will :-)

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -465,29 +486,122 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+		config_flags |= LL_TIM_IC_POLARITY_RISING;
+	}
+	/* TODO: add 'vendor' config flags for filter, prescaler div */
+	config_flags |= (LL_TIM_ICPSC_DIV1 | LL_TIM_IC_FILTER_FDIV1);
```
Can remove the parentheses pair.

XenuIsWatching replied:
done

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -511,7 +630,7 @@ static DEVICE_API(counter, counter_stm32_driver_api) = {
 				LL_TIM_ClearFlag_CC##cc(timer);			\
 			}							\
-			counter_stm32_alarm_irq_handle(dev, cc - 1U);		\
+			counter_stm32_irq_handle(dev, cc - 1U);		\
```
Could you add an extra tab?

XenuIsWatching replied:
done

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -53,6 +54,12 @@ static uint32_t(*const get_timer_compare[TIMER_MAX_CH])(TIM_TypeDef *) = {
 };
 #endif
+#ifdef CONFIG_COUNTER_CAPTURE
+static uint32_t (*const get_timer_capture[TIMER_MAX_CH])(const TIM_TypeDef *) = {
```
```c
/** Channel to timestamp capture function mapping. */
```
Could you add an empty line before `#ifdef ...` above and after `#endif` below?

XenuIsWatching replied:
done

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -465,29 +486,122 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	/* With Counter Capture, we need to check which mode it was configured for */
+	if (LL_TIM_IC_GetActiveInput(timer, LL_TIM_CHANNEL_CH1 << (4 * id)) ==
+	    LL_TIM_ACTIVEINPUT_DIRECTTI) {
+		counter_capture_cb_t cb;
+
+		cb = chdata->capture;
```
```suggestion
		counter_capture_cb_t cb = chdata->capture;

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -465,29 +486,122 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	const struct counter_stm32_config *config = dev->config;
+	struct counter_stm32_ch_data *chdata = &config->ch_data[chan];
+	TIM_TypeDef *timer = config->timer;
+	uint32_t config_flags = LL_TIM_ACTIVEINPUT_DIRECTTI;
```
I think it would be worth the call `disable_it[chan](timer);` here.

XenuIsWatching replied:
it shouldn't be disabling the interrupt in the case where it's being used as a counter capture, the alarm is a one-shot alarm just disables it in that case.

etienne-lms replied:
I was thinking of preventing the interrupt to triggers in between `chdata->capture` and `chdata->user_data` update right below.

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -113,6 +113,35 @@ extern "C" {
+
+ /**@} */
+
+typedef void (*counter_capture_cb_t)(const struct device *dev, uint8_t chan_id,
```
Suggestion
```
/** @brief Counter capture callback
 *
 * @param dev       Pointer to the device structure for the driver instance
 * @param chan_id   Channel ID
 * @param flags     Configuration flags (COUNTER_CAPTURE_*)
 * @param ticks     Counter value that triggered the capture
 * @param user_data User data
 */

XenuIsWatching replied:
done

etienne-lms replied:
CI compliance check taints on the space char missing before `*` character above.

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -304,6 +345,92 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
+	const struct counter_driver_api *api =
+				(struct counter_driver_api *)dev->api;
+
+	if (!api->capture_callback_set) {
```
```suggestion
	if (api->capture_callback_set == NULL) {
```
Ditto for `z_impl_counter_capture_enable()` and `z_impl_counter_capture_disable()`.

etienne-lms replied:
Well, the whol file should be updated. Zephyr coding guideline explicitly requires comparison against NULL here.

XenuIsWatching replied:
done

etienne-lms reviewed (commented):
In the commit message, would you mind replacing `apis` with `APIs`.

etienne-lms commented on include/zephyr/drivers/counter.h:
```diff
@@ -113,6 +113,43 @@ extern "C" {
+ *
+ * @param dev       Pointer to the device structure for the driver instance
+ * @param chan_id   Channel ID
+ * @param flags     Configuration flags (COUNTER_CAPTURE_*)
```
Since there is an *anchor* for flags:
```suggestion
 * @param flags     Configuration flags (@ref COUNTER_CAPTURE_FLAGS)
```
Ditto for `counter_capture_callback_set()` inline description comment.

lindblandro commented on include/zephyr/drivers/counter.h:
```diff
@@ -304,6 +353,92 @@ static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
+ * @retval 0 If successful.
+ * @retval Negative error code on failure
+ */
+__syscall int counter_capture_callback_set(const struct device *dev,
```
The [syscall docs](https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html) mention 

>APIs that allow the user to register callback functions that run in supervisor mode should never be exposed as system calls. Reserve these for supervisor-mode access only.

Doesn't this allow exactly that?

etienne-lms commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -465,29 +488,121 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+		config_flags |= LL_TIM_IC_POLARITY_BOTHEDGE;
+	} else if (flags & COUNTER_CAPTURE_FALLING_EDGE) {
+		config_flags |= LL_TIM_IC_POLARITY_FALLING;
+	} else if (flags & COUNTER_CAPTURE_RISING_EDGE) {
```
```suggestion
	} else if (flags & COUNTER_CAPTURE_RISING_EDGE != 0) {
```

Alternatively:
```c
	switch (flags & COUNTER_CAPTURE_BOTH_EDGES) {
	case COUNTER_CAPTURE_BOTH_EDGES:
		config_flags |= LL_TIM_IC_POLARITY_BOTHEDGE;
		break;
	case COUNTER_CAPTURE_FALLING_EDGE:
		config_flags |= LL_TIM_IC_POLARITY_FALLING;
		break;
	case COUNTER_CAPTURE_RISING_EDGE:
		config_flags |= LL_TIM_IC_POLARITY_RISING;
		break;
	default:
		break;	
	}

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -494,28 +528,200 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+		LOG_ERR("%s: Capture channel %d is enabled, cannot reconfigure", dev->name, chan);
+		return -EBUSY;
+	}
+
```
yeah...i suppose this is a good check (why am i talking to a 🤖 )

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -411,6 +437,14 @@ static int counter_stm32_init_timer(const struct device *dev)
+	r = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (r < 0) {
+		LOG_ERR("%s: Counter Capture pinctrl setup failed (%d)", dev->name, r);
+		return r;
```
hmm... good point

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -1206,6 +1329,104 @@ static inline int z_impl_counter_set_value_64(const struct device *dev, uint64_t
+ *
+ * @retval 0 If successful.
+ * @retval Negative error code on failure
+ */
```
this really depends on the SoC... just at least the STM32, can't be configured while it is enabled. it's really implementation defined and the api shouldn't be forced for this, will not do

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -1206,6 +1329,104 @@ static inline int z_impl_counter_set_value_64(const struct device *dev, uint64_t
+	}
+
+	return api->capture_callback_set_64(dev, chan_id, flags, cb, user_data);
+}
```
yep, will do

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -494,28 +528,200 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+				flags = COUNTER_CAPTURE_FALLING_EDGE;
+			} else if (pol == LL_TIM_IC_POLARITY_BOTHEDGE) {
+				flags = COUNTER_CAPTURE_BOTH_EDGES;
+			}
```
this is inaccurate, if both edges are enabled, it will return which one happened (rising OR falling)

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -494,28 +531,223 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	const struct counter_stm32_config *config = dev->config;
+	TIM_TypeDef *timer = config->timer;
+
+	LL_TIM_CC_DisableChannel(timer, LL_TIM_CHANNEL_CH1 << (4 * chan));
```
no, this is by design, you shouldn't have to call configure before every enable

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -484,29 +537,230 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	chdata->single_shot_capture = (flags & COUNTER_CAPTURE_SINGLE_SHOT) ? true : false;
+
+	/* Apply configuration */
+	LL_TIM_IC_Config(timer, LL_TIM_CHANNEL_CH1 << (4 * chan), config_flags);
```
yep... but it is to be ch2ll[chan - 1]

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -484,29 +537,230 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	const struct counter_stm32_config *config = dev->config;
+	TIM_TypeDef *timer = config->timer;
+
+	LL_TIM_CC_DisableChannel(timer, ch2ll[chan - 1]);
```
yep

XenuIsWatching commented on drivers/counter/counter_stm32_timer.c:
```diff
@@ -484,29 +537,230 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
+	const struct counter_stm32_config *config = dev->config;
+	TIM_TypeDef *timer = config->timer;
+
+	LL_TIM_CC_DisableChannel(timer, ch2ll[chan - 1]);
```
done

ubieda reviewed (commented):
Thanks for this PR - I stumbled upon it as I'm interested in using this peripheral functionality with Zephyr.

I left some comments and I'll try it out on a Nucleo.

XenuIsWatching commented:
> Thanks for this PR - I stumbled upon it as I'm interested in using this peripheral functionality with Zephyr.
> 
> I left some comments and I'll try it out on a Nucleo.

Thanks!!! There's more work on this on top this where I'm working on a counter-cap-dev branch in my fork which will allow for more tighter integration with existing drivers such as sensors where it can use this counter capture along with its hardware timestamp for the trigger rather than a gpio.

teburd commented:
> > Thanks for this PR - I stumbled upon it as I'm interested in using this peripheral functionality with Zephyr.
> > I left some comments and I'll try it out on a Nucleo.
> 
> Thanks!!! There's more work on this on top this where I'm working on a counter-cap-dev branch in my fork which will allow for more tighter integration with existing drivers such as sensors where it can use this counter capture along with its hardware timestamp for the trigger rather than a gpio.

Sounds cool!

ubieda reviewed (dismissed):
Tested on a Nucleo G071RB and it works as expected.

Approving assuming the concerns on the API are eased by marking this feature as Experimental.

XenuIsWatching commented on drivers/counter/Kconfig:
```diff
@@ -10,6 +10,14 @@ menuconfig COUNTER
 
+config COUNTER_CAPTURE
+	bool "Counter capture"
+	help
```
done

XenuIsWatching commented on include/zephyr/drivers/counter.h:
done

XenuIsWatching commented on include/zephyr/drivers/counter.h:
```diff
@@ -1206,6 +1340,120 @@ static inline int z_impl_counter_set_value_64(const struct device *dev, uint64_t
+
+	return api->capture_disable(dev, chan_id);
+}
+#endif /* CONFIG_COUNTER_CAPTURE */
```
will be done in another PR... code is in development
https://github.com/XenuIsWatching/stm32-h5-zephyr/blob/master/src/capture_test.c

XenuIsWatching committed:
drivers: counter: stm32: introduce counter capture api

This introduces APIs for input capture on the STM32.

Rename callback in data to alarm_cb to prevent confusion with other
callbacks.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: introduce counter capture api

This introduces APIs for input capture.

A use case for this would be time stamping a rising and/or falling edge
where precision and lack of jitter is important. The only way to currently
do this in zephyr is to set up a GPIO interrupt and then read the counter
or kernel time... but this can have limitations of software such as the
processor may have a lot going on and can be subject to interrupt
latencies. Having a hardware 'latch' done of the rising and/or falling
edge with respect to a timer can eliminate these imprecisions.

How this is to be used, is that an application is to call
`capture_callback_set` with the configuration of the channel number, flags
(such as falling, rising, or both edges), and then a callback function.
Then the `capture_enable` function is to be called to enable this. The
callback has the timestamp of the edge in ticks and the edge that it
captured (rising or falling).

This also adds a 64 bit tick version for the capture callback.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
Changes in latest push
* fix function names in `counter_handlers.c`
* split commit in to 2; one for the api additions and one for implementing it in stm32

XenuIsWatching commented:
merge conflicts fixed in latest push

nordic-krch commented on drivers/counter/Kconfig:
```diff
@@ -10,6 +10,15 @@ menuconfig COUNTER
+	  hardware timestamping of external events as well as a software
+	  interrupt on the event. This feature is not supported by all
+	  counter drivers.
+
```
what about adding Kconfig that indicates whether counter implementation supports the capture feature (`COUNTER_SUPPORT_CAPTURE`?). Something like async API in UART driver: https://github.com/zephyrproject-rtos/zephyr/blob/acc53653b4707aa40c22aec8c33f1d733e5e670f/drivers/serial/Kconfig#L22.

XenuIsWatching replied:
make sense, will do

XenuIsWatching replied:
done, `COUNTER_SUPPORTS_CAPTURE` added

nordic-krch commented on include/zephyr/drivers/counter.h:
```diff
@@ -1289,6 +1397,132 @@ static inline int z_impl_counter_set_value_64(const struct device *dev, uint64_t
+ * @retval 0 If successful.
+ * @retval Negative error code on failure
+ */
+static inline int counter_capture_callback_set(const struct device *dev, uint8_t chan_id,
```
I've seen many times a case where getting interrupt when capture event occur was not expected. On the other hand, capture value was done later, after the operation completed. In other words, I'm missing a function for reading the captured value and not enabling interrupt.

What about renaming it to `counter_capture_configure()` (allow null callback) and adding `counter_capture_get(dev, chan, &value)`

XenuIsWatching replied:
I'm not so sure about this... although i've heard similiar feedback like this before.

My thoughts about this are how would you know with `counter_capture_get` if a capture really occured or if you were reading the last capture again (it wouldn't be easy to just check the ts as it wraps due to overflow)

althrougth that would be fairly easy to implement to just 'cache' the timestamp in to driver data and then just return it (along with a status code (and would still require an interrupt to happen within the driver)), i'm not sure what the pratical use of that would be. I don't quite understand what you mean there about "I've seen many times a case where getting interrupt when capture event occur was not expected. On the other hand, capture value was done later, after the operation completed. In other words, I'm missing a function for reading the captured value and not enabling interrupt."
i figured with the isr callback, the application could rather do whatever it wants to handle it in it's own way, weither store in to an array, or just 'toss' it

nordic-krch replied:
I'm probably biased because Nordic TIMER just don't have the interrupt on capture. In case of capturing a GPIO event, interrupt would come from GPIO driver anyway.

As for use cases, one example: Let say you want to capture the width of the short pulse. One channel captures rising edge, another falling. If pulse is short there is no point of getting interrupt on the first slope.

Anyway, having configure function which allows null is just more flexible as you can have the same behavior as with `capture_callback_set` but you can have interrupt-less version with `capture_get`. The responsibility about validity of the capture is outside of the counter driver in that case.

XenuIsWatching replied:
> I'm probably biased because Nordic TIMER just don't have the interrupt on capture. In case of capturing a GPIO event, interrupt would come from GPIO driver anyway.

But Nordic has the peripheral interconnect, right? You would just use a GPIOTE + PPI, right? and you just interrupt on the timer compare interrupt. [I may be wrong here, it's been a while since i've looked at the databook here]

Anyways, I'm probably okay with renaming the `capture_callback_set()` to `counter_capture_configure()`, i just would much rather defer this `capture_get` api to come with a use case when/if that is implemented. We have this marked as `experimental` anyway here

XenuIsWatching commented:
change in the last push
* rename `counter_capture_callback_set` to `counter_capture_configure`

nordic-krch commented on include/zephyr/drivers/counter.h:
```diff
@@ -1289,6 +1397,132 @@ static inline int z_impl_counter_set_value_64(const struct device *dev, uint64_t
+
+/**
+ * @brief Set callback function for a counter timestamp capture
+ *
```
API does not specify which pin is captured. I guess that it's ok as some SoC may have fixed connections but maybe it's worth mentioning that GPIO setup is vendor specific and done in DT or out of bounds. Otherwise it looks incomplete as there are flags to pick the edge of the capture but pin in nowhere mentioned.

XenuIsWatching replied:
this is already handled by the 'chan', the SoC (stm32) that we have just implemented just has fixed connections it's a direct connection to specific pads which is done by the DT which is the pinmux in this case

I think that is a good point and maybe there could be a `source` argument added or a `counter_capture_set_source` api with it in the future.... but I think we should cross that bridge when we get there. right now with the only driver i have implemented there wouldn't have a use for that api or argument

I can just add a comment that the channel to source pairing is vendor specific and done in DT for now

XenuIsWatching commented:
Change in last push
* doxygen updated on `counter_capture_configure` and `counter_capture_configure_64`

towen commented:
Thanks @XenuIsWatching - I hope this can be merged soon :rocket: 
I have tested your driver with the following code, which could perhaps be added as a sample
<details> <summary>main.c  (expand)</summary>

```
#include <zephyr/kernel.h>
#include <zephyr/drivers/counter.h>

#define CAPTURE_CHANNEL_ID 0
#define CAPTURE_FLAGS (COUNTER_CAPTURE_CONTINUOUS | COUNTER_CAPTURE_FALLING_EDGE)

#define SAMPLE_TIMER DT_INST(0, st_stm32_counter)


static void counter_capture_cb(const struct device *counter_dev, uint8_t chan_id,
				     counter_capture_flags_t flags, uint32_t cap_ticks,
				     void *user_data)
{
	uint32_t now_ticks;
	uint64_t now_nsec, cap_nsec;

	counter_get_value(counter_dev, &now_ticks);

	now_nsec = counter_ticks_to_ns(counter_dev, now_ticks);
	cap_nsec = counter_ticks_to_ns(counter_dev, cap_ticks);

	printk("Interrupt latency %lld ns \n", now_nsec-cap_nsec);
}

int main(void)
{
	const struct device *const counter_dev = DEVICE_DT_GET(SAMPLE_TIMER);
	int err;

	printk("Counter capture sample\n\n");

	if (!device_is_ready(counter_dev)) {
		printk("device not ready.\n");
		return 0;
	}

	counter_start(counter_dev);

	err = counter_capture_configure(counter_dev, CAPTURE_CHANNEL_ID, CAPTURE_FLAGS, &counter_capture_cb, NULL);
	err |= counter_enable_capture(counter_dev, CAPTURE_CHANNEL_ID);

	if (err != 0) {
		printk("Error %d\n", err);
	}
	else {
		printk("Counter capture armed\n");
	}

	while (1) {
		k_sleep(K_FOREVER);
	}
	return 0;
}
```

</details>

Touching the PA0 pin of my nucleo_g474re triggers the counter capture, which measures Zephyr's interrupt latency (in its default configuration on stm32) at around 3950 ns.

XenuIsWatching commented:
> Thanks @XenuIsWatching - I hope this can be merged soon 🚀 I have tested your driver with the following code, which could perhaps be added as a sample

yeah... it could, maybe in other pr after this lands, i do have a tester code in another pr here https://github.com/zephyrproject-rtos/zephyr/pull/104446

nordic-krch reviewed (approved):
API looks good.
